### PR TITLE
feat: automatically select LoRA modules when none are provided

### PR DESCRIPTION
**Pull request by RobotSail**
In the current version of the training library, we have the default value of target_modules set to
a list oflayer names which are implementation-specific and may not reflect what a given model actually
uses for the layer names. Furthermore, the default is also a subset of all projection layers in most models,
and the recommendation is generally to use all of these layers when injecting low rank adapters.

This commit resolves that issue by introducing logic to automatically resolve the target modules
and default to using all of them when they are not provided. This commit also adds validation logic
which indicates when some of the provided modules do not exist in the model. To go a step further,
the training library will also now error out when none of the provided target modules exist in the model,
supplying the user with additional context on which modules exist and how they could resolve the error

Signed-off-by: Oleg S <ec2-user@ip-10-0-24-47.us-east-2.compute.internal>

### CODE REVIEW

**Comment by RobotSail** [comment]
resolves #164

File: src/instructlab/training/utils.py [hunk 1]
```diff
@@ -733,3 +735,11 @@ def set_random_seed(seed):
         np.random.seed(seed)
         torch.manual_seed(seed)
         torch.cuda.manual_seed_all(seed)
+
+
+def get_projection_layer_names(model: PreTrainedModel) -> List[str]:
+    """
+    Given a pretrained model, returns all of the projection layers (matching '_proj')
+    """
+    proj_layers = set(name.split('.')[-1] for name, _ in model.named_modules() if name.endswith("_proj"))
```
[the remark this answers is not in the document]
Comment (fabianlim) [reply src/instructlab/training/utils.py:744]: another alternative if you want to target all the linears, is to do `isinstance (mod, torch.nn.Linear)`

[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/training/utils.py:744]: @fabianlim The models we actively support are listed here:
https://github.com/instructlab/training/blob/9e2ac746a877e2b2ed6ff2ba54a04c1a22dadb84/src/instructlab/training/main_ds.py#L158

When I looked at their list of layers, all of them had `k_proj, q_proj, v_proj, o_proj`, so the assumption is true at least for supported models. 

You're right though, we could go down the path of targeting all linear layers. I just have two questions about this approach:

1. How would this affect the memory requirements?
2. What would be the impact on training times? If we are targeting more modules for LoRA, we would potentially be dropping more pretrained weights in favor of our LoRA approximations - how would this impact the loss curve?

[the remark this answers is not in the document]
Comment (fabianlim) [reply src/instructlab/training/utils.py:744]: If it is as you said. That you are targeting all the proj, then it should be equivalent to putting a Lora adapter on all linears

[the remark this answers is not in the document]
Comment (RobotSail) [reply src/instructlab/training/utils.py:744]: @fabianlim Not necessarily. Some models will use Linear layers which are not explicitly labeled as projections. For example, in starcoder-3b, these account roughly 1.2B parameters: 

Screenshot 2024-08-19 at 1 16 02 PM

**Review by Maxusmusti — APPROVED** [review]
LGTM

File: src/instructlab/training/main_ds.py [hunk 2]
```diff
@@ -682,7 +699,8 @@ def run_training(torch_args: TorchrunArgs, train_args: TrainingArgs) -> None:
                 "--lora_target_modules",
             ]
         )
-        command.extend(train_args.lora.target_modules)
+        if train_args.lora.target_modules:
+            command.extend(train_args.lora.target_modules)
```
Comment (aldopareja) [inline src/instructlab/training/main_ds.py:703]: should we have this only for granite models?, how about non-granite models?

**Review by aldopareja — APPROVED** [review]
I think this look alright, just a minor comment that can be ignored for the moment.

**Review by JamesKunstle — APPROVED** [review]
lgtm

**Comment by JamesKunstle** [comment]
@RobotSail Is this good to go post-rebase?

**Comment by RobotSail** [comment]
@JamesKunstle Yeah it should be. Mergify didn't merge it automatically

File: src/instructlab/training/main_ds.py (hunk 2 above)
Comment (RobotSail) [reply src/instructlab/training/main_ds.py:722]: Specifying the target modules? It should be fine for all models, since we may want to target different modules depending on what we want them to learn
